exercises/dnd-character/src/test/java/DnDCharacterTest.java

```diff
@@ -128,7 +130,144 @@ public void testRandomCharacterIsValid() {
     @Ignore("Remove to run test")
     @Test
     public void testEachAbilityIsOnlyCalculatedOnce() {
-        assertEquals(dndCharacter.getStrength(), dndCharacter.getStrength());
+        int initialStrength     = dndCharacter.getStrength();
+        int initialDexterity    = dndCharacter.getDexterity();
+        int initialConstitution = dndCharacter.getConstitution();
+        int initialIntelligence = dndCharacter.getIntelligence();
+        int initialWisdom       = dndCharacter.getWisdom();
+        int initialCharisma     = dndCharacter.getCharisma();
+        for (int j = 0; j < 100; j++) {
+            assertEquals(initialStrength,     dndCharacter.getStrength());
+            assertEquals(initialDexterity,    dndCharacter.getDexterity());
+            assertEquals(initialConstitution, dndCharacter.getConstitution());
+            assertEquals(initialIntelligence, dndCharacter.getIntelligence());
+            assertEquals(initialWisdom,       dndCharacter.getWisdom());
+            assertEquals(initialCharisma,     dndCharacter.getCharisma());
+        }
+    }
+
+    @Ignore("Remove to run test")
+    @Test
+    public void testAbilityIsProbablyNotConstant() {
+        int ability = dndCharacter.ability();
+        for (int i = 0; i < 100; i++) {
+            if (ability != dndCharacter.ability()) {
+                return;
+            }
+        }
+        fail();
     }
 
+    @Ignore("Remove to run test")
+    @Test
+    public void testAbilityIsDistributedAccordingToRules() {
+        /* This test is meant to check if the 'ability' method is actually
+         * returning a (pseudo-)random number which respects the rules of the
+         * game:
+         *     Roll four dice. Take the sum of the largest three dice.
+         *
+         * This test was added to filter out two common, wrong implementations:
+         *     Returning a random number between 3 and 18.
+         *     Returning the sum of three random numbers each between 1 and 6.
+         *
+         * The way we check if the 'ability' method is distributed correctly is
+         * by creating a histogram of the distribution by calling ability
+         * repeatedly. Then we compare it to the reference distribution within
+         * a certain tolerance.
+         */
+
+        // These two values were chosen empirically.
```

Comment:
I'm not sure empirically chosen values are going to be sufficient and may lead to test flakiness. Running this test on the reference solution needs to consistently pass (pretty much every time since we gate PRs on passing all tests). Can you make sure that this at least passes on at least 1000 runs of this test?

Another option is - can you modify the reference solution to have a test-only constructor which fixes the Random object with a specific seed that definitely passes the tests?

That is, instead of creating a Random object [here](https://github.com/exercism/java/blob/master/exercises/dnd-character/.meta/src/reference/java/DnDCharacter.java#L67) which will use time to help seed the object, we set it in the constructor and make a test-only constructor that takes in the Random object so that we can seed it specifically for this test case. Thoughts?

Reply:
@jmrunkle: The test is flaky, but at an astronomically small rate.

If we fix the seed, the next progression is this:

![random_number](https://user-images.githubusercontent.com/50879/79585706-7b652300-80d0-11ea-860d-737b88b3915e.png)

The novel approach here, which is not something you see in a lot of places, is to make a probabilistic property test: Every time you run the test suite, you strenghten your belief, and the chance of a false positive is quantifiably insignificant. It'd be neat if it said in a comment exactly / approximately what the odds are, except, that would put a lot of strain on the contributor.

Reply:
Fixing a seed value would not actually break the testing logic at all. It just guarantees that your test *always* passes. Rather than having someone quantify the "astronomically small rate" why don't we just make our setup resilient to test failures?

I'm suggesting having a setup like this where the object takes a seed (or even the Random object itself) in the constructor. Then we have some margin of control over the randomness:

```java
private final Random random;

DnDCharacter() {
  this(new Random());
}

/** test-only constructor used for ensuring stability of tests results */
DnDCharacter(Random random) {
  this.random = random;
}
```

Most of the tests would still be able to use the empty constructor, but this one test would be able to take advantage of a fixed seed to ensure that our tests on the reference solution always pass. The only cost is that some "astronomically unlikely" student might have a test that always fails because it just happens to fail that test with the seed we chose. I much prefer that.

Reply:
> Fixing a seed value would not actually break the testing logic at all.

Not fixing a seed value would not actually break the testing logic, either.

> Rather than having someone quantify the "astronomically small rate" why don't we just make our setup resilient to test failures?

Because property-based testing with arbitrary, random variables means that running the unit test again and again tests the property from new angles every single time.

> I'm suggesting having a setup like this where the object takes a seed (or even the Random object itself) in the constructor.

This is a very good suggestion. I'm saying this because that's exactly what I suggested in #1671. ;-) This was downvoted by two people about a year ago. Perhaps it is better to keep the two subjects of probabilistic testing and dependency injection separate since they're both being contended?

Reply:
> This is a very good suggestion. I'm saying this because that's exactly what I suggested in #1671. ;-) This was downvoted by two people about a year ago. Perhaps it is better to keep the two subjects of probabilistic testing and dependency injection separate since they're both being contended?

I would have argued in favor of this back when that debate was active. If you want to see the real effects of this thing on repeated trials, you can simply modify the test locally to not set a seed on the Random object. Automated tests that we rely on for actual development should pass, reliably, every time. I honestly do not accept any other opinion on that (especially when there are a lot of tests being run at once).

As such, I guess I am really saying that we either (1) need to add the dependency injection or (2) get rid of a distribution test that might be flaky. I prefer adding DI, but maybe that's just me.

Back in #1671, it was argued that this is only a "level 2" difficulty exercise. Maybe it should have the difficulty set a bit higher if we think DI is too difficult for early exercises?

Reply:
> we either (1) need to add the dependency injection or (2) get rid of a distribution test that might be flaky

Why? The distribution test currently supports the global state of `Random`'s side-effects.

It's not flaky in practice. You can decide exactly how often you will get a false negative and make that frequency so large we won't experience it in our lifetimes. This cannot mean the same as if the frequency were observable.

While I'm a big fan of DI, it isn't necessary to make a probabilistic test. In fact, if the exercise had DI, this would instead encourage mock testing, which is argued in #1671.

At this point I sense "too many chefs", so I'll let you decide and will not object to any decision made.

Reply:
> this would instead encourage mock testing

I think that's highly debatable and well within our control as maintainers. It's not easy to mock the Random object precisely because you have no idea how they will use it. I just want to fix the seed so that we can guarantee that the test passes consistently for our reference implementation.

Reply:
> Can you make sure that this at least passes on at least 1000 runs of this test?

My solution has passed the updated test 1000000 times without any failure.

> this one test would be able to take advantage of a fixed seed to ensure that our tests on the reference solution always pass. The only cost is that some "astronomically unlikely" student might have a test that always fails because it just happens to fail that test with the seed we chose. I much prefer that.

I disagree. If the test should fail for the "astronomically unlikely" student with a conforming solution, they can just rerun the tests, which then should succeed, as well as all following runs. Further there may also be an "astronomically unlikely" false positive solution for that seed, which would consistently pass the tests.

The way we circumvent these issues is precisely by having the `Random` Object being seeded with a different seed every time.

As the solution has already passed the test 1000000 times, I think it is safe to assume that it will continue to do so for the foreseeable future. If there are still concerns about this I might try to figure out the actual probability of this test failing on a conforming solution (as well as passing a non-conforming one).

If we choose a fixed seed we can only control our reference solution to consistently pass the tests. Not the solutions of our students. Thus if we want to make this exercise enjoyable to them, we must hold ourselves to the same standard as them.